Client/Experiments/Messaging/GleanPlumbMessageManager.swift

```diff
@@ -88,39 +90,78 @@ class GleanPlumbMessageManager: GleanPlumbMessageManagerProtocol {
-            .filter {
-                $0.data.surface == surface
-            }
+        return getNextMessage(for: surface, availableMessages: getMessages(messagingFeature.value()))
+    }
 
+    func getNextMessage(for surface: MessageSurfaceId, availableMessages: [GleanPlumbMessage]) -> GleanPlumbMessage? {
         // If `NimbusMessagingHelper` creation fails, we cannot continue with this
         // feature! For that reason, return `nil`. We need to recreate the helper
         // for each request to get a message because device context can change.
         guard let messagingHelper = helperUtility.createNimbusMessagingHelper() else { return nil }
+        var jexlCache = [String: Bool]()
 
-        // Take the first triggered message.
-        guard let message = getNextTriggeredMessage(messages, messagingHelper) else { return nil }
+        var excluded: Set<String> = []
+        return getNextMessage(for: surface,
+                              availableMessages: availableMessages,
+                              excluded: &excluded,
+                              messagingHelper: messagingHelper,
+                              jexlCache: &jexlCache)
+    }
+
+    func getNextMessage(
+            for surface: MessageSurfaceId,
+            availableMessages: [GleanPlumbMessage],
+            excluded: inout Set<String>,
+            messagingHelper: NimbusMessagingHelperProtocol,
+            jexlCache: inout [String: Bool]
+    ) -> GleanPlumbMessage? {
+        let feature = messagingFeature.value()
+        let message = availableMessages
+            .first { message in
+                guard message.data.surface == surface &&
+                        !excluded.contains(message.id) else {
+                    return false
+                }
+                do {
+                    return try evaluationUtility.isMessageEligible(message,
+                                                                   messageHelper: messagingHelper,
+                                                                   jexlCache: &jexlCache)
+                } catch {
+                    return false
+                }
+            }
+        guard let message = message else {
+            return nil
+        }
 
         // If it's a message under experiment, we need to react to whether it's a control or not.
-        if message.isUnderExperimentWith(key: feature.messageUnderExperiment) {
-            guard let nextTriggeredMessage = handleMessageUnderExperiment(message,
-                                                                          messages,
-                                                                          messagingHelper,
-                                                                          feature.onControl)
-            else { return nil }
-
-            return nextTriggeredMessage
+        guard let slug = message.data.experiment else {
+            return message
         }
```

Comment:
```suggestion
        guard let slug = message.data.experiment else { return message }
```
nit: guard styles. We don't have a linter for this yet, because I have to write a custom rule for it and haven't had a chance. :(